### PR TITLE
8284105: Update security libraries to use sealed classes

### PR DESCRIPTION
**Pull request by seanjmullan**
Please review these changes to update the security libraries to use sealed classes. See JEP 409 for more details on sealed classes.

No CSR is required as all the changes are to internal classes. A few classes that did not have subclasses were simply marked final instead of sealed.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8284105](https://bugs.openjdk.java.net/browse/JDK-8284105): Update security libraries to use sealed classes

### Reviewers
 * [Joe Darcy](https://openjdk.java.net/census#darcy) (@jddarcy - **Reviewer**) ⚠️ Review applies to 52aa0ad6bdf099cf6cd422bca214d1d9720e7e0a
 * [Weijun Wang](https://openjdk.java.net/census#weijun) (@wangweij - **Reviewer**) ⚠️ Review applies to c27b1ba02742a17bf67b56e63651e997c6b59327
 * [Xue-Lei Andrew Fan](https://openjdk.java.net/census#xuelei) (@XueleiFan - **Reviewer**) ⚠️ Review applies to 52aa0ad6bdf099cf6cd422bca214d1d9720e7e0a

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/8165/head:pull/8165` \
`$ git checkout pull/8165`

Update a local copy of the PR: \
`$ git checkout pull/8165` \
`$ git pull https://git.openjdk.java.net/jdk pull/8165/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8165`

View PR using the GUI difftool: \
`$ git pr show -t 8165`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/8165.diff">https://git.openjdk.java.net/jdk/pull/8165.diff</a>

</details>

### CODE REVIEW

**Review by jddarcy — APPROVED** [review]

File: src/java.base/share/classes/sun/security/util/math/intpoly/IntegerPolynomial.java [hunk 1]
```diff
@@ -61,7 +61,14 @@
  *
  */
 
-public abstract class IntegerPolynomial implements IntegerFieldModuloP {
```
Comment (wangweij) [inline src/java.base/share/classes/sun/security/util/math/intpoly/IntegerPolynomial.java:64]: Although we only have several implementations, I think this class is meant to be freely extendable for whatever new modulus.

File: src/java.base/share/classes/sun/security/rsa/RSASignature.java [hunk 2]
```diff
@@ -47,7 +47,7 @@
  * @since   1.5
  * @author  Andreas Sterbenz
  */
-public abstract class RSASignature extends SignatureSpi {
```
Comment (wangweij) [inline src/java.base/share/classes/sun/security/rsa/RSASignature.java:50]: We can probably move the `RSASignature.encodeSignature` method to `RSAUtil` and this class can be package private.

**Review by wangweij — COMMENTED** [review]
It looks `KrbTgsRep.java`, `Krb5ProxyCredential.java`, `Builder.java`, `Vertex.java`, `Validator.java`, and `RSAKeyPairGenerator.java` can all be made package private.

**Review by XueleiFan — APPROVED** [review]
It looks safe to me as if compiling and test passed.

**Comment by seanjmullan** [comment]
> It looks `KrbTgsRep.java`, `Krb5ProxyCredential.java`, `Builder.java`, `Vertex.java`, `Validator.java`, and `RSAKeyPairGenerator.java` can all be made package private.

Good point, although I would prefer to leave `Validator` as public for now until we are more sure of the compatibility risk as there have been external dependencies on it.

It may be useful to apply the `sealed` modifier to package-private classes, but for this RFE that is not the goal, so I will remove the `sealed` modifier from these classes (if applicable) when I make them package-private.

File: src/java.base/share/classes/sun/security/util/math/intpoly/IntegerPolynomial.java (hunk 1 above)
Comment (seanjmullan) [reply src/java.base/share/classes/sun/security/util/math/intpoly/IntegerPolynomial.java:64]: We can always add new ones to the permits clause later.

File: src/java.security.jgss/share/classes/sun/security/krb5/KrbTgsRep.java [hunk 3]
```diff
@@ -40,7 +40,7 @@
  * This class encapsulates a TGS-REP that is sent from the KDC to the
  * Kerberos client.
  */
-public class KrbTgsRep extends KrbKdcRep {
+final class KrbTgsRep extends KrbKdcRep {
```
Comment (wangweij) [inline src/java.security.jgss/share/classes/sun/security/krb5/KrbTgsRep.java:43]: Make `KrbAsRep` final also to be symmetric.

**Review by wangweij — APPROVED** [review]
Only one comment. Others look fine.

**Pull request closed without merging**